clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/mapreduce/datawriter/reduce/VeniceReducer.java

```diff
@@ -0,0 +1,235 @@
+import org.apache.hadoop.mapred.Counters;
+import org.apache.hadoop.mapred.JobClient;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.JobID;
+import org.apache.hadoop.mapred.OutputCollector;
+import org.apache.hadoop.mapred.Reducer;
+import org.apache.hadoop.mapred.Reporter;
+import org.apache.hadoop.mapred.RunningJob;
+import org.apache.hadoop.util.Progressable;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * {@link VeniceReducer} will be in charge of producing the messages to Kafka broker.
+ * Since {@link VeniceMRPartitioner} is using the same logic of {@link com.linkedin.venice.partitioner.DefaultVenicePartitioner},
+ * all the messages in the same reducer belongs to the same topic partition.
+ *
+ * The reason to introduce a reduce phase is that BDB-JE will benefit with sorted input in the following ways:
+ * 1. BDB-JE won't generate so many BINDelta since it won't touch a lot of BINs at a time;
+ * 2. The overall BDB-JE insert rate will improve a lot since the disk usage will be reduced a lot (BINDelta will be
+ * much smaller than before);
+ *
+ */
+public class VeniceReducer extends AbstractPartitionWriter
+    implements Reducer<BytesWritable, BytesWritable, BytesWritable, BytesWritable> {
+  public static final String MAP_REDUCE_JOB_ID_PROP = "mapreduce.job.id";
+  private static final Logger LOGGER = LogManager.getLogger(VeniceReducer.class);
+
+  private VeniceProperties props;
+
+  private Reporter reporter = null;
+  private DataWriterTaskTracker dataWriterTaskTracker = null;
+  private HadoopJobClientProvider hadoopJobClientProvider = new DefaultHadoopJobClientProvider();
+
+  private MapReduceEngineTaskConfigProvider engineTaskConfigProvider;
+
+  @Override
+  protected EngineTaskConfigProvider getEngineTaskConfigProvider() {
+    return engineTaskConfigProvider;
+  }
+
+  protected JobConf getJobConf() {
+    return engineTaskConfigProvider.getJobConf();
+  }
+
+  @Override
+  public void configure(JobConf job) {
+    this.engineTaskConfigProvider = new MapReduceEngineTaskConfigProvider(job);
+    super.configure();
+  }
+
+  @Override
+  public void reduce(
+      BytesWritable key,
+      Iterator<BytesWritable> values,
+      OutputCollector<BytesWritable, BytesWritable> output,
+      Reporter reporter) {
+    if (updatePreviousReporter(reporter)) {
+      setCallback(new ReducerProduceCallback(reporter));
+      dataWriterTaskTracker = new ReporterBackedMapReduceDataWriterTaskTracker(reporter);
+    }
+    processValuesForKey(key.copyBytes(), mapIterator(values, BytesWritable::copyBytes), dataWriterTaskTracker);
+  }
+
+  private boolean updatePreviousReporter(Reporter reporter) {
+    if (this.reporter == null || !this.reporter.equals(reporter)) {
+      this.reporter = reporter;
+      return true;
+    }
+    return false;
+  }
+
+  // Visible for testing
+  @Override
+  protected boolean getExceedQuotaFlag() {
```

Comment:
nit: If it is for unit test, all the "protected" can be removed.

Reply:
The problem here is with the structure of tests - I did not add tests specific to `AbstractPartitionWriter`, and so, it falls back to the existing tests for `VeniceReducer`. However, since most of the core functionality has moved to `AbstractPartitionWriter`, there is a chaining of function calls to set these configs.

`AbstractPartitionWriter` and `VeniceReducer` are in different packages and we will not be able to invoke the methods in `AbstractPartitionWriter` if the access specifier is not `protected`